### ISSUE TITLE
EF 7 SQL statements with no filter are generated for LINQ query using FirstOrDefault()

### ISSUE THREAD

**Issue by jemiller0** [aspnet/EntityFramework#3807]
The following code queries for blogs and returns the first post for the blog. I checked the SQL that is executed using SQL Server Profiler. First a query for the blogs table is executed. That's fine. Then, it appears to execute a separate query for each blog against the posts table. It appears that this is wrong though. The posts queries don't have a where clause. It should be limiting those to just the blog at hand. The end result comes out correct. I'm assuming filtering is happening in memory. However, if the posts table contains lots of rows, it would be unacceptably slow. If it's doing N+1 queries, it should including a WHERE clause and filtering by id in the main table.

```c#
using Microsoft.Data.Entity;
using System.Collections.Generic;
using System.Linq;

namespace ConsoleApplication1
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var bc = new BlogContext())
            {
                bc.Database.EnsureCreated();
                for (int i = 0; i < 10; i++)
                {
                    var b = new Blog { };
                    bc.Blogs.Add(b);
                    var p = new Post { Blog = b };
                    bc.Posts.Add(p);
                }
                bc.SaveChanges();
                bc.Blogs
                    //.Take(100)
                    //.Select(b2 => b2.Posts.Any())
                    .Select(b2 => b2.Posts.FirstOrDefault().Content)
                    .ToArray();
            }
        }
    }

    public class BlogContext : DbContext
    {
        public DbSet<Blog> Blogs { get; set; }
        public DbSet<Post> Posts { get; set; }
        //protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) => optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=Blogs;Trusted_Connection=True;");
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) => optionsBuilder.UseSqlServer(@"Server=localhost;Database=Blogs;Trusted_Connection=True;");
    }

    public class Blog
    {
        public int? Id { get; set; }
        public ICollection<Post> Posts { get; set; }
    }

    public class Post
    {
        public int? Id { get; set; }
        public string Content { get; set; }
        public int? BlogId { get; set; }
        public Blog Blog { get; set; }
    }
}
```
```SQL
SELECT [b2].[Id]
FROM [Blog] AS [b2]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
SELECT [p].[BlogId], [p].[Content]
FROM [Post] AS [p]
go
```

**Comment by toddlucas** [comment]
I'm seeing a similar issue with rc1-final with the npgsql provider. I'm doing ```context.TableSet.Where(t => t.col1 == x && t.col2 == y)```. The logged SQL does not have a where clause, regardless of the query forcing function (ToListAsync, FirstOrDefaultAsync, etc.). I added an index and the abysmal performance continued, so I drilled down and found this.

**Comment by divega** [comment]
@toddlucas if your query actually looks like that and no WHERE clause is generated it is most likely a different root causr. Please file a separate issue and include a repro.

**Comment by rowanmiller** [comment]
@maumar do you know if this is still n+1

**Comment by maumar** [comment]
@rowanmiller yes, checked on latest bits

**Comment by anpete** [comment]
@maumar Do the inner queries have a where clause?

**Comment by rowanmiller** [comment]
@maumar ping 😄

**Comment by rowanmiller** [comment]
@maumar ping again 😄

**Comment by maumar** [comment]
@anpete I copied the initial repro and got the same result as the customer. Now I also added filter:

```
bc.Blogs
    .Select(b2 => b2.Posts.Where(pp => pp.BlogId > 0).FirstOrDefault().Content)
    .ToArray()
```

And the query plan is as follows:

```
TRACKED: False
(QueryContext queryContext) => IEnumerable<string> _Select(
    source: IEnumerable<ValueBuffer> _ShapedQuery(
        queryContext: queryContext, 
        shaperCommandContext: SelectExpression: 
            SELECT [b2].[Id]
            FROM [Blogs] AS [b2]
        , 
        shaper: ValueBufferShaper
    )
    , 
    selector: (ValueBuffer b2) => string FirstOrDefault(
        source: IEnumerable<string> _Select(
            source: IEnumerable<ValueBuffer> _Where(
                source: IEnumerable<ValueBuffer> _ShapedQuery(
                    queryContext: queryContext, 
                    shaperCommandContext: SelectExpression: 
                        SELECT [pp].[BlogId], [pp].[Content]
                        FROM [Posts] AS [pp]
                        WHERE [pp].[BlogId] > 0
                    , 
                    shaper: ValueBufferShaper
                )
                , 
                predicate: (ValueBuffer pp) => (Nullable<int>) object b2.get_Item(0) == (Nullable<int>) object pp.get_Item(0)
            )
            , 
            selector: (ValueBuffer pp) => (string) object pp.get_Item(1)
        )
    )
)

```

**Comment by maumar** [comment]
poaching per discussion with @anpete

**Comment by maumar** [comment]
fixed in 24c3db2c3c1d711d9d424addc52f82574def4e65
